src/par_iter/find_first_last.rs

```diff
@@ -0,0 +1,183 @@
+use std::cell::Cell;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use super::internal::*;
+use super::*;
+use super::len::*;
+
+// The consumer for find_first/find_last has fake indexes representing the lower
+// and upper bounds of the "range" of data it consumes. This range does not
+// correspond to indexes from the consumed iterator but rather indicate the
+// consumer's position relative to other consumers. The purpose is to allow a
+// consumer to know it should stop consuming items when another consumer finds a
+// better match.
+
+// An indexed consumer could specialize to use the real indexes instead, but we
+// don't implement that for now. The only downside of the current approach is
+// that in some cases, iterators very close to each other will have the same
+// range and therefore not be able to stop processing if one of them finds a
+// better match than the others.
```

Comment:
> I think AtomicU64 is not stable yet (tracked here). Would it be worth postponing merging this PR until it's stable? Of course, we should be able to make that change later, because the exact size of the integer used is private to this module.

Nah, just use a usize. It'll make it easier to write a test. :)

As for the test... well, you could... perhaps write something that drives the splitting by hand? i.e., take a u64 range, create a producer from it, and manually call split etc? The test can be internal to this module if it needs access to private state...

...at minimum some thorough comments for how that case is handled would be a good start.